addons/pos_loyalty/static/src/overrides/models/pos_store.js

```diff
@@ -234,56 +226,33 @@ patch(PosStore.prototype, {
         const domain = new Domain(reward_product_domain);
 
         try {
-            products
-                .filter((product) => domain.contains(product))
-                .forEach((product) => reward.all_discount_product_ids.add(product.id));
+            for (const p of products) {
+                const serializedProduct = p.serialize();
+
+                if (domain.contains(serializedProduct)) {
+                    reward.all_discount_product_ids.add(p.id);
+                }
+            }
         } catch (error) {
             if (!(error instanceof InvalidDomainError)) {
                 throw error;
             }
-            const index = this.rewards.indexOf(reward);
+            const index = this.models["loyalty.reward"].indexOf(reward);
```

Comment:
if `reward` is already the record, you don't need to look for it. Use it directly. So in the line below where `description` is accessed, you can just `reward.description`.